### ISSUE TITLE
Differentiate between user cancel and deleted keychain item in VALSecureEnclaveValet.

### ISSUE THREAD

**Issue by nijindal** [square/Valet#67]
I am using `VALSecureEnclaveValet` for storing some sensitive data.

My app needs to behave differently for use-case when user cancels the popup and when user disables PIN and then re-enables it.

Since VALSecureEnclaveValet's `objectForKey:userPrompt:` returns nil for both the cases, how do I differentiate?

**Comment by dfed** [comment]
Use `containsObjectForKey:`. This will not prompt the customer for Touch ID or device passcode, and will tell you if an object exists in the keychain. If it exists in the keychain and you get back `nil`, it means the customer cancelled.

Closing this as resolved, but do let me know if this doesn't answer your question!

**Comment by dfed** [comment]
@EricMuller22, we might want to add `containsObjectForKey:` to `VALSecureEnclaveValet`'s header with some documentation that it doesn't prompt the customer. This is the second question of this type we've fielded in a week... Thoughts on whether a little documentation would help?

**Comment by nijindal** [comment]
@dfed thanks for the quick response. That worked.

But I think there is a bug in `containsObjectForKey:` implementation.

Steps to reproduce:
- Add a key in `VALSecureEnclaveValet`
- Go into Setting -> Disable Passcode
- Re-Enable Passcode
- Go back to the app and query for the stored key
- Now, `containsObjectForKey:` object still returns `YES` whereas that key was deleted from the keychain

Now this case will be inferred like a user-cancel by the app whereas the value is actually deleted.

**Comment by dfed** [comment]
Interesting!! Re-opening to investigate. I'm reasonably sure that's a bug in Keychain, but I'll take a look. Good find!

**Comment by dfed** [comment]
@nijindal what operating system(s) did you run that test on?

**Comment by nijindal** [comment]
I tried it with apple sample for touchID keychain demo. It gives correct
error, "item not found".

On Mon, 18 Jan 2016 at 9:22 am, Dan Federman <notifications@github.com>
wrote:

> Interesting!! Re-opening to investigate. I'm reasonably sure that's a bug
> in Keychain, but I'll take a look. Good find!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/square/Valet/issues/67#issuecomment-172418431>.
>
-- 
Jindal

**Comment by nijindal** [comment]
iPhone 6. iOS 8.4

On Mon, 18 Jan 2016 at 9:25 am, Dan Federman <notifications@github.com>
wrote:

> @nijindal <https://github.com/nijindal> what operating system(s) did you
> run that test on?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/square/Valet/issues/67#issuecomment-172418616>.
>
-- 
Jindal

**Comment by dfed** [comment]
@nijindal can you link me to the apple code you were using? If we're returning `YES`, that means that we got `errSecSuccess` or `errSecInteractionNotAllowed`. Weird that you got `errSecItemNotFound`... if Valet were seeing that it would have returned `NO`. I'll investigate more this week, but that's seriously curious – looking at the sample code you were using would help.

**Comment by nijindal** [comment]
https://developer.apple.com/library/ios/samplecode/KeychainTouchID/Introduction/Intro.html

Try running this piece with same steps. It will return the error.

**Comment by dfed** [comment]
I assume you're looking at `copyMatchingAsync` and seeing the message created on line 70 of  `AAPLKeychainTestsViewController.m`. My initial guess is that Keychain tells them `ItemNotFound` because they have user authentication (i.e. Touch ID) to check the secure enclave. `VALSecureEnclaveValet`'s `containsObjectForKey:` does not use customer authentication, so it looks like we're just being told the wrong thing by Keychain in this case. Investigating.

**Comment by dfed** [comment]
Confirmed that the keychain is giving us the wrong information when we don't prompt for user interaction in this one case. That's supremely annoying.

**Comment by nicwise** [comment]
Hi! I've just been trying this, too, as I need it for my app.

I see you are using iOS8.4. iOS8 doesn't support the "did the finger prints change" thing. Did 8 support it when the pin was turned off? What param are you using to create the valet @nijindal ?

I tried it on 9.2.1, and I did this:

1. Setup the valet using 

```
self.secureEnclaveValet = [[VALSecureEnclaveValet alloc] initWithIdentifier:@"UserPresence" accessControl:VALAccessControlTouchIDCurrentFingerprintSet];
```
2. Create the item
3. Turn off passcode
4. Try to read it:

```
NSString *password = [self.secureEnclaveValet stringForKey:self.username userPrompt:@"Use TouchID to retreive password"];
    
    if (password) {
        self.textView.text = [self.textView.text stringByAppendingFormat:@"\n%s - we got %@ back - all good", __PRETTY_FUNCTION__, password];
    } else {
        if ([self.secureEnclaveValet containsObjectForKey:self.username]) {
            self.textView.text = [self.textView.text stringByAppendingFormat:@"\n%skey is there, so user pressed cancel", __PRETTY_FUNCTION__];
        } else {
            self.textView.text = [self.textView.text stringByAppendingFormat:@"\n%s key is not there - fingerprints changed", __PRETTY_FUNCTION__];
        }
    }
```

This shows "finger prints changed" - even if I turn it back on with the same fingerprint set.

Even when I turn the passcode back on, the item is still not there.

(I'm using the updated branch tho - https://github.com/square/Valet/tree/federman/secaccesscontrolflags)

There was also a [bug in iOS8.3 which might be relevant](https://forums.developer.apple.com/thread/3607)

**Comment by dfed** [comment]
I'm strongly considering adding something like the following API to all Valet types:
`objectForKey:containsObjectForKey:` `stringForKey:containsObjectForKey:` where the second parameter is an in/out BOOL pointer. I'll also be filing a radar about the issue where we get `InteractionNotAllowed` instead of `ItemNotFound` after a customer removes the passcode on their device and then adds one back.

@nicwise, @nijindal is using a vanilla `VALSecureEnclaveValet` (which uses `kSecAccessControlUserPresence` under the hood).

**Comment by nicwise** [comment]
Oh for Swift's tuples :)

**Comment by dfed** [comment]
When we go to 3.0 / Swift, you can bet your arse we'll be using a tuple there.

**Comment by nicwise** [comment]
Or throw an exception. that could work too.....​ depends on your view on
exceptions.

**Comment by dfed** [comment]
My view is: throw exceptions only when there's absolutely no way to recover from what just happened. So that's unlikely to be where we end up with that one :)

**Comment by nicwise** [comment]
I'd agree in ObjC, but my background is C#, where exceptions are cheap and used for flow control a lot (as they are in Java). I suspect Swift might be the same.... we'll see as it happens over time.

Tuple would work too....

**Comment by dfed** [comment]
Went with a more targeted fix. Should be exactly what you're looking for. Let me know what you think!

**Comment by nijindal** [comment]
Looks good to me.

**Comment by dfed** [comment]
Just filed rdar://24283957

**Comment by dfed** [comment]
Valet v2.2.0 is now out with this fix.

**Comment by nijindal** [comment]
Thanks for the quick fix @dfed 

I was trying this out and I came across another unexpected behaviour. 

If I make two subsequent calls to `VALSecureEnclaveValet` for fetching a key which is stored
- cancel the first authentication popup
- Subsequent fetch call doesn't trigger an authentication pop up and gives an `errSecAuthFailed` error.

Similarly, keychain gives `errSecAuthFailed` in case you lock the phone when popup is on screen. 

Since userCancelled comes out to be false in both the cases above, app tends to believe that data has been deleted.

I have attached a sample project demonstrating the first one.
[keychain.zip](https://github.com/square/Valet/files/101936/keychain.zip)

**Comment by dfed** [comment]
Interesting! Thanks for finding all the good bugs 😃

Sounds like we need to add `errSecAuthFailed` to our user cancelled check. @nijindal want to give us a PR? We'll roll to v2.2.1. Otherwise I can take care of this sometime next week

**Comment by nijindal** [comment]
@dfed I was pondering over the real use-cases of `errSecAuthFailed`. These were some cases I found while doing basic testing. But merging all those use cases to "user cancel" doesn't sound right. 

Documentation just says __The user name or passphrase you entered is not correct.__ which doesn't suffice for all the cases. Any comments ?

**Comment by nijindal** [comment]
I was going through the list of errors possible while trying to access the value. So assuming `nil` return && "non user cancel" to be item deleted can be a wrong assumption in other cases too. e.g., `errSecAllocate`. 

Some of these might be temporary errors but assuming them to be data deletion might be catastrophic for some apps. Like in my app, if I know that data stored was deleted, I just reset the whole app. So, I want stronger confirmation on data deletion. I think `errSecItemNotFound` should also be exposed in the API. What do you think ?

**Comment by dfed** [comment]
hrm. I wouldn't worry about `errSecAllocate`, since the likelihood that you're app is still running under that scenario is insanely low (you'll more likely have crashed before that happens).

We could expose ItemNotFound in the API, thought naming the parameter `itemNotFound` makes for some less-than-easy reading. I'd likely want to name this in the positive: `itemFound`. This would be set to true if `status != errSecItemNotFound && status != errSecNotAvailable && status != errSecInteractionNotAllowed && status != errSecUserCanceled`. I think that would do what you want, right?

The odd part is whether to include `UserCanceled` in that check. I like having the `userCancelled:` flag and the `itemFound:` flag be mutually exclusive. You're reset code check would need to check `(!userCancelled && !itemFound)`, which seems reasonable. Thoughts?

cc @EricMuller22

**Comment by nijindal** [comment]
Sounds good. I also prefer the exclusivity of both the flags.

**Comment by nijindal** [comment]
Hey @dfed
Are you planning to make this change anytime soon ?

I am kind of waiting on it and doesn't have bandwidth right now to open a PR for it.

**Comment by dfed** [comment]
@nijindal I'll get to this issue this week. Filing a new issue so we can move discussion to an open issue.

**Comment by nijindal** [comment]
@dfed I have added a comment on PR.
